README.md
A small PHP library to obfuscate numbers. Use it when you don't want to expose your database numeric ids to users.

The permutations are collision-free and fully deterministic. The random-looking effect is due to encryption, not to a PRNG. The same boundary range with the same secret key will always produce the same output.

[![Latest Version on Packagist](https://img.shields.io/packagist/v/northlands/permuteseq.svg?style=flat-square)](https://packagist.org/packages/northlands/permuteseq)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/northlands-io/permuteseq/run-tests?label=tests)](https://github.com/northlands-io/permuteseq/actions?query=workflow%3Arun-tests+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/northlands/permuteseq.svg?style=flat-square)](https://packagist.org/packages/northlands/permuteseq)

## Installation

You can install the package via composer:

```bash
composer require northlands/permuteseq
```

## Usage

```php
$permuteseq = new Permuteseq(123456789012345, 1000, 9999); // Range 1000-9999

$permuteseq = Permuteseq::fixed(123456789012345, 1000, 9999);
$permuteseq = Permuteseq::dynamic(123456789012345, 8);

$encoded = $permuteseq->encode(1000); // 4070
$decoded = $permuteseq->decode($encoded); // 1000
```

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Disclaimer

The library should not be considered cryptographically strong or used to sensitive data.

## Credits

* https://github.com/dverite/permuteseq \
The PostgreSQL extension by Daniel Vérité.

* https://czep.net/21/obfuscate.html \
An article on how to obfuscate primary keys in databases.

## Alternatives

* https://github.com/ioleo/cryptomute \
Another Format Preserving Encryption library supporting multiple ciphers (DES, AES and Camellia) but slower performance.

* https://github.com/vinkla/hashids \
Generates YouTube-like ids from numbers. Can be limited to digit-alphabet, but returns strings like `"09284"` instead of safe integers.

* https://github.com/jenssegers/optimus \
Super-fast number obfuscation based on Knuth's integer hash. However, range can only be defined by 4-62 bits.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v0.2 - 2022-09-05

Support PHP 8.0 versions.

## v0.1 - 2022-09-05

First release. 🎉


tests/PermuteseqTest.php
<?php

it('can encode and decode random numbers', function () {
    $permuteseq = new Northlands\Permuteseq\Permuteseq(123456789012345, 0, 1000);

    for ($number = 0; $number < 1000; $number++) {
        $encoded = $permuteseq->encode($number);

        expect($number)->toEqual($permuteseq->decode($encoded));
    }
});

it('is collision-free', function () {
    $permuteseq = new Northlands\Permuteseq\Permuteseq(123456789012345, 0, 1000);

    $unique = [];

    for ($number = 0; $number < 1000; $number++) {
        $encoded = $permuteseq->encode($number);

        expect(array_key_exists($encoded, $unique))->toBeFalse();

        $unique[$encoded] = true;
    }
});

it('supports static constructor', function () {
    $permuteseq = Northlands\Permuteseq\Permuteseq::create(123456789012345);

    expect($permuteseq)->toBeInstanceOf(Northlands\Permuteseq\Permuteseq::class);
});

it('throws exception if not 64-bit secret key', function () {
    new Northlands\Permuteseq\Permuteseq(1234);
})->throws(InvalidArgumentException::class, "Key must be 64-bit integer.");

it('throws exception with invalid range arguments', function () {
    new Northlands\Permuteseq\Permuteseq(123456789012345, 1, 3);
})->throws(InvalidArgumentException::class, "Invalid range");

it('throws exception with invalid rounds argument', function () {
    new Northlands\Permuteseq\Permuteseq(123456789012345, 0, 1000, 2);
})->throws(InvalidArgumentException::class, "Must be an odd integer greater or equal to 3.");

it('throws exception when input is out of range', function () {
    $permuteseq = new Northlands\Permuteseq\Permuteseq(123456789012345, 1000, 9999);

    $permuteseq->encode(10000);
})->throws(InvalidArgumentException::class, "Value out of range.");


src/Permutation.php
<?php

namespace App\Support;

use InvalidArgumentException;

/**
 * @see https://github.com/dverite/permuteseq/blob/dcff6e91eaebe69e3ee9b77e3cf00d6f66edb425/permuteseq.c
 */
class Permutation
{
    protected $key;

    protected $min;

    protected $max;

    protected $rounds;

    /**
     * @param int $key
     * @param int $min
     * @param int $max
     * @param int $rounds Number of rounds of the Feistel Network. Must be an odd integer greater or equal to 3.
     */
    public function __construct(int $key, int $min = 0, int $max = PHP_INT_MAX, int $rounds = 7)
    {
        if (! $this->hasValidRange($min, $max)) {
            throw new InvalidArgumentException("Invalid range: The difference between minimum and maximum values should be at least 3.");
        }

        if ($rounds < 3) {
            throw new InvalidArgumentException("Must be an odd integer greater or equal to 3.");
        }

        $this->key = $key;
        $this->min = $min;
        $this->max = $max;
        $this->rounds = $rounds;
    }

    public static function create(int $key, int $min = 0, int $max = PHP_INT_MAX, int $rounds = 7): static
    {
        return new static($key, $min, $max, $rounds);
    }

    /**
     * The output is constrained to the boundaries of the range by
     * using a cycle-walking cipher on top of a Feistel network.
     *
     * @param $value
     * @param bool $reverse
     * @return int
     */
    public function permute($value, bool $reverse = false): int
    {
        if ($value < $this->min || $value > $this->max) {
            throw new InvalidArgumentException("Value out of range.");
        }

        /**
         * Arbitrary maximum number of "walks" along the results
         * searching for a value inside the [min,max] range.
         * It's mainly to avoid an infinite loop in case the chain of
         * results has a cycle (which would imply a bug somewhere).
         */
        $max = 1000000;

        $interval = $this->max - $this->min + 1;

        $count = 0;

        /**
         * Compute the half block size: it's the smallest power of 2 such as two
         * blocks are greater than or equal to the size of interval in bits. The
         * half-blocks have equal lengths.
         */
        $hsz = 1;

        while ($hsz < 32 && 1 << (2*$hsz) < $interval) {
            $hsz++;
        }

        $mask = (1 << $hsz) - 1;

        /**
         * Scramble the key. This is not strictly necessary, but will
         * help if the user-supplied key is weak, for instance with only a
         * few right-most bits set.
         */
        $key = $this->hash($this->key & 0xffffffff) | $this->hash(($this->key >> 32) & 0xffffffff) << 32;

        /**
         * Initialize the two half blocks.
         * Work with the offset into the interval rather than the actual value.
         * This allows to use the full 32-bit range.
         */
        $l1 = ($value - $this->min) >> $hsz;
        $r1 = ($value - $this->min) & $mask;

        $l2 = $r2 = 0;

        do {
            for ($i = 0; $i < $this->rounds; $i++) {
                $l2 = $r1;

                /**
                 * The sub-key Ki for the round i is a sliding and cycling window
                 * of hsz bits over K, moving left to right, so each round takes
                 * different bits out of the crypt key. The round function is
                 * simply hash(Ri) XOR hash(Ki).
                 * When decrypting, Ki corresponds to the Kj of encryption with
                 * j=(NR-1-i), i.e. we iterate over sub-keys in the reverse order.
                 */
                $ki = $key >> ($hsz * ($reverse ? $this->rounds - 1 - $i : $i) & 0x3f);
                $ki += $reverse ? $this->rounds - 1 - $i : $i;

                $r2 = ($l1 ^ $this->hash($r1) ^ $this->hash($ki)) & $mask;

                $l1 = $l2;
                $r1 = $r2;
            }

            $result = ($r1 << $hsz) | $l1;

            $l1 = $r2;
            $r1 = $l2;
        } while ($result > $this->max - $this->min && $count++ < $max);

        if ($count >= $max) {
            throw new \RuntimeException("Infinite cycle walking detected.");
        }

        // Convert the offset in the interval to an absolute value, possibly negative.
        return $this->min + $result;
    }

    public function encode($value): int
    {
        return $this->permute($value);
    }

    public function decode($value): int
    {
        return $this->permute($value, true);
    }

    protected function hasValidRange($min, $max): bool
    {
        if (($min > 0 && $max < PHP_INT_MIN + $min) || ($min < 0 && $max > PHP_INT_MAX + $min)) {
            return true;
        }

        return $max - $min >= 4 - 1;
    }

    /**
     * @see https://doxygen.postgresql.org/hashfn_8c.html#a0e8a5084b019b55453fa64ac0329e73e
     * @see https://doxygen.postgresql.org/hashfn_8c.html#ab4646d77540701d2eb2c877effbe5739
     */
    protected function hash($k): int
    {
        $a = $b = $c = 0x9e3779b9 + 4 + 3923095;

        $a += $k;

        $c ^= $b; $c -= $this->rot($b, 14);
        $a ^= $c; $a -= $this->rot($c, 11);
        $b ^= $a; $b -= $this->rot($a, 25);
        $c ^= $b; $c -= $this->rot($b, 16);
        $a ^= $c; $a -= $this->rot($c, 4);
        $b ^= $a; $b -= $this->rot($a, 14);
        $c ^= $b; $c -= $this->rot($b, 24);

        return (int) $c;
    }

    /**
     * @see https://doxygen.postgresql.org/hashfn_8c.html#aae44e21ada356d9d84450d5440fbb0c4
     */
    protected function rot($x, $k): int
    {
        return ($x << $k) | ($x >> (32 - $k));
    }
}


src/BasePermutation.php
<?php

namespace Northlands\Permuteseq;

use InvalidArgumentException;

abstract class BasePermutation implements PermutationInterface
{
    protected int $key;

    protected int $min;

    protected int $max;

    protected int $rounds;

    /*public function __construct(int $key, int $rounds = 7)
    {
        $this->key = $key;
        $this->rounds = $rounds;
    }*/

    public function permute($value, bool $reverse = false): int
    {
        if ($value < $this->min || $value > $this->max) {
            throw new InvalidArgumentException("Value out of range.");
        }

        /**
         * Arbitrary maximum number of "walks" along the results
         * searching for a value inside the [min,max] range.
         * It's mainly to avoid an infinite loop in case the chain of
         * results has a cycle (which would imply a bug somewhere).
         */
        $max = 1000000;

        $interval = $this->max - $this->min + 1;

        $count = 0;

        /**
         * Compute the half block size: it's the smallest power of 2 such as two
         * blocks are greater than or equal to the size of interval in bits. The
         * half-blocks have equal lengths.
         */
        $hsz = 1;

        while ($hsz < 32 && 1 << (2*$hsz) < $interval) {
            $hsz++;
        }

        $mask = (1 << $hsz) - 1;

        /**
         * Scramble the key. This is not strictly necessary, but will
         * help if the user-supplied key is weak, for instance with only a
         * few right-most bits set.
         */
        $key = $this->hash($this->key & 0xffffffff) | ($this->hash($this->key >> 32) & 0xffffffff) << 32;

        /**
         * Initialize the two half blocks.
         * Work with the offset into the interval rather than the actual value.
         * This allows to use the full 32-bit range.
         */
        $l1 = ($value - $this->min) >> $hsz;
        $r1 = ($value - $this->min) & $mask;

        $l2 = $r2 = 0;

        do {
            for ($i = 0; $i < $this->rounds; $i++) {
                $l2 = $r1;

                /**
                 * The sub-key Ki for the round i is a sliding and cycling window
                 * of hsz bits over K, moving left to right, so each round takes
                 * different bits out of the crypt key. The round function is
                 * simply hash(Ri) XOR hash(Ki).
                 * When decrypting, Ki corresponds to the Kj of encryption with
                 * j=(NR-1-i), i.e. we iterate over sub-keys in the reverse order.
                 */
                $ki = $key >> ($hsz * ($reverse ? $this->rounds - 1 - $i : $i) & 0x3f);
                $ki += ($reverse ? $this->rounds - 1 - $i : $i);

                $r2 = ($l1 ^ $this->hash($r1) ^ $this->hash($ki)) & $mask;

                $l1 = $l2;
                $r1 = $r2;
            }

            $result = ($r1 << $hsz) | $l1;

            // Swap one more time to prepare for the next cycle
            $l1 = $r2;
            $r1 = $l2;
        } while (($result < 0 || $result > $this->max - $this->min) && $count++ < $max);

        if ($count >= $max) {
            throw new \RuntimeException("Infinite cycle walking detected.");
        }

        // Convert the offset in the interval to an absolute value, possibly negative.
        return $this->min + $result;
    }

    /**
     * @see https://doxygen.postgresql.org/hashfn_8c.html#a0e8a5084b019b55453fa64ac0329e73e
     * @see https://doxygen.postgresql.org/hashfn_8c.html#ab4646d77540701d2eb2c877effbe5739
     */
    protected function hash($k): int
    {
        $a = $b = $c = 0x9e3779b9 + 4 + 3923095;

        $a += $k;

        $c ^= $b; $c -= $this->rot($b, 14);
        $c &= 0xffffffff;

        $a ^= $c; $a -= $this->rot($c, 11);
        $a &= 0xffffffff;

        $b ^= $a; $b -= $this->rot($a, 25);
        $b &= 0xffffffff;

        $c ^= $b; $c -= $this->rot($b, 16);
        $c &= 0xffffffff;

        $a ^= $c; $a -= $this->rot($c, 4);
        $a &= 0xffffffff;

        $b ^= $a; $b -= $this->rot($a, 14);
        $b &= 0xffffffff;

        $c ^= $b; $c -= $this->rot($b, 24);

        return (int) $c & 0xffffffff;
    }

    /**
     * @see https://doxygen.postgresql.org/hashfn_8c.html#aae44e21ada356d9d84450d5440fbb0c4
     */
    protected function rot($x, $k): int
    {
        return ($x << $k) | ($x >> (32 - $k));
    }

    protected function hasValidRange($min, $max): bool
    {
        if (($min > 0 && $max < PHP_INT_MIN + $min) || ($min < 0 && $max > PHP_INT_MAX + $min)) {
            return true;
        }

        return $max - $min >= 4 - 1;
    }
}


src/DynamicPermutation.php
<?php

namespace Northlands\Permuteseq;

use InvalidArgumentException;

class DynamicPermutation extends BasePermutation
{
    protected int $minlength;

    protected int $maxlength;

    public function __construct(int $minlength, int $maxlength = null, int $rounds = 7)
    {
        if ($minlength < 1) {
            throw new InvalidArgumentException("Minimum length must be 1 or greater.");
        }

        $this->minlength = $minlength;
        $this->maxlength = $maxlength;

        // TODO The $maxlength behave like an upper limit that won't expand.

        $this->rounds = $rounds;
    }

    public function permute($value, bool $reverse = false): int
    {
        if ($value < 0) {
            throw new InvalidArgumentException("Number must be positive.");
        }
        // TODO Adjust min/max based on input length

        $length = $value > 9 ? floor(log10($value + 1)) : 1;

        if ($length < $this->minlength) {
            $length = $this->minlength;
        }

        $this->min = pow(10, $length);
        $this->max = $this->min * 10 - 1;

        if (! $this->hasValidRange($this->min, $this->max)) {
            throw new InvalidArgumentException("Invalid range: The difference between minimum and maximum values should be at least 3.");
        }

        if ($reverse) {
            $value -= $this->min; // TODO Post permute?
        } else {
            $value += $this->min;
        }

        return parent::permute($value, $reverse);
    }
}


src/Permuteseq.php
<?php

namespace Northlands\Permuteseq;

use InvalidArgumentException;

/**
 * @see https://github.com/dverite/permuteseq/blob/dcff6e91eaebe69e3ee9b77e3cf00d6f66edb425/permuteseq.c
 */
class Permuteseq
{
    protected int $key;

    protected int $min;

    protected int $max;

    protected int $rounds;

    /**
     * @param int $key
     * @param int $min
     * @param int $max
     * @param int $rounds Number of rounds of the Feistel Network. Must be an odd integer greater or equal to 3.
     */
    final public function __construct(int $key, int $min = 0, int $max = PHP_INT_MAX, int $rounds = 7)
    {
        if ($key < 2_147_483_648) {
            throw new InvalidArgumentException("Key must be 64-bit integer.");
        }

        if (! $this->hasValidRange($min, $max)) {
            throw new InvalidArgumentException("Invalid range: The difference between minimum and maximum values should be at least 3.");
        }

        if ($rounds < 3) {
            throw new InvalidArgumentException("Must be an odd integer greater or equal to 3.");
        }

        /**
         * Scramble the key. This is not strictly necessary, but will
         * help if the user-supplied key is weak, for instance with only a
         * few right-most bits set.
         *
         * The dverite/permuteseq extension supports 0xFFFFFFFFFFFFFFFF.
         * This project cannot be fully compatible without GMP or BCMath extensions.
         */
        $key = ($this->hash($key & 0xFFFFFFFF) | ($this->hash($key >> 32) & 0xFFFFFFFF) << 32) & 0x7FFFFFFFFFFFFFFF;

        $this->key = $key;
        $this->min = $min;
        $this->max = $max;
        $this->rounds = $rounds;
    }

    public static function create(int $key, int $min = 0, int $max = PHP_INT_MAX, int $rounds = 7): static
    {
        return new static($key, $min, $max, $rounds);
    }

    /**
     * The output is constrained to the boundaries of the range by
     * using a cycle-walking cipher on top of a Feistel network.
     *
     * @param $value
     * @param bool $reverse
     * @return int
     */
    public function permute($value, bool $reverse = false): int
    {
        if ($value < $this->min || $value > $this->max) {
            throw new InvalidArgumentException("Value out of range.");
        }

        /**
         * Arbitrary maximum number of "walks" along the results
         * searching for a value inside the [min,max] range.
         * It's mainly to avoid an infinite loop in case the chain of
         * results has a cycle (which would imply a bug somewhere).
         */
        $max = 1000000;

        $interval = $this->max - $this->min + 1;

        $count = 0;

        /**
         * Compute the half block size: it's the smallest power of 2 such as two
         * blocks are greater than or equal to the size of interval in bits. The
         * half-blocks have equal lengths.
         */
        $hsz = 1;

        while ($hsz < 32 && 1 << (2*$hsz) < $interval) {
            $hsz++;
        }

        $mask = (1 << $hsz) - 1;

        /**
         * Initialize the two half blocks.
         * Work with the offset into the interval rather than the actual value.
         * This allows to use the full 32-bit range.
         */
        $l1 = ($value - $this->min) >> $hsz;
        $r1 = ($value - $this->min) & $mask;

        $l2 = $r2 = 0;

        do {
            for ($i = 0; $i < $this->rounds; $i++) {
                $l2 = $r1;

                /**
                 * The sub-key Ki for the round i is a sliding and cycling window
                 * of hsz bits over K, moving left to right, so each round takes
                 * different bits out of the crypt key. The round function is
                 * simply hash(Ri) XOR hash(Ki).
                 * When decrypting, Ki corresponds to the Kj of encryption with
                 * j=(NR-1-i), i.e. we iterate over sub-keys in the reverse order.
                 */
                $ki = $this->key >> ($hsz * ($reverse ? $this->rounds - 1 - $i : $i) & 0x3f);
                $ki &= 0xFFFFFFFF;

                $ki += ($reverse ? $this->rounds - 1 - $i : $i);
                $ki &= 0xFFFFFFFF;

                $r2 = ($l1 ^ $this->hash($r1) ^ $this->hash($ki)) & $mask;

                $l1 = $l2;
                $r1 = $r2;
            }

            $result = ($r1 << $hsz) | $l1;

            // Swap one more time to prepare for the next cycle
            $l1 = $r2;
            $r1 = $l2;
        } while (($result < 0 || $result > $this->max - $this->min) && $count++ < $max);

        if ($count >= $max) {
            throw new \RuntimeException("Infinite cycle walking detected.");
        }

        // Convert the offset in the interval to an absolute value, possibly negative.
        return $this->min + $result;
    }

    public function encode($value): int
    {
        return $this->permute($value);
    }

    public function decode($value): int
    {
        return $this->permute($value, true);
    }

    protected function hasValidRange($min, $max): bool
    {
        if (($min > 0 && $max < PHP_INT_MIN + $min) || ($min < 0 && $max > PHP_INT_MAX + $min)) {
            return true;
        }

        return $max - $min >= 4 - 1;
    }

    /**
     * @see https://doxygen.postgresql.org/hashfn_8c.html#a0e8a5084b019b55453fa64ac0329e73e
     * @see https://doxygen.postgresql.org/hashfn_8c.html#ab4646d77540701d2eb2c877effbe5739
     */
    protected function hash($k): int
    {
        $a = $b = $c = 0x9e3779b9 + 4 + 3923095;

        $a += $k;

        $c ^= $b; $c -= $this->rot($b, 14);
        $c &= 0xFFFFFFFF;

        $a ^= $c; $a -= $this->rot($c, 11);
        $a &= 0xFFFFFFFF;

        $b ^= $a; $b -= $this->rot($a, 25);
        $b &= 0xFFFFFFFF;

        $c ^= $b; $c -= $this->rot($b, 16);
        $c &= 0xFFFFFFFF;

        $a ^= $c; $a -= $this->rot($c, 4);
        $a &= 0xFFFFFFFF;

        $b ^= $a; $b -= $this->rot($a, 14);
        $b &= 0xFFFFFFFF;

        $c ^= $b; $c -= $this->rot($b, 24);
        $c &= 0xFFFFFFFF;

        return (int) $c;
    }

    /**
     * @see https://doxygen.postgresql.org/hashfn_8c.html#aae44e21ada356d9d84450d5440fbb0c4
     */
    protected function rot($x, $k): int
    {
        return ($x << $k) | ($x >> (32 - $k));
    }
}
